doc/development/adding_templates.rst

```diff
@@ -0,0 +1,91 @@
+
+PennyLane internally distinguishes :ref:`different types <intro_ref_temp>` of templates, such as
+:mod:`embeddings <pennylane.templates.embeddings>`, :mod:`layers <pennylane.templates.layers>`,
+:mod:`state_preparations <pennylane.templates.state_preparations>` or
+:mod:`soubroutines <pennylane.templates.soubroutines>`.
+Here we will use the template ``MyNewTemplate`` as an example, and you need to replace ``<templ_type>`` with the
+correct template type.
+
+1. **Add the template** by adding a new file ``my_new_template.py`` to the correct ``templates/<templ_type>/``
+   subdirectory, which contains your new template of the form
+
+   .. code-block:: python
+
+        @template
+        def MyNewTemplate(*args, **kwargs):
+            ...
+
+   A template is a function that defines a sequence of quantum gates (without measurements).
+   Since the template is called within a :ref:`quantum function <intro_vcirc_qfunc>`,
+   it can only contain information processing that is allowed
+   inside a quantum functions.
+
+   Make sure you consider the following:
+
+   * To be consistent with other quantum operations (which are classes), the function name (i.e., ``MyNewTemplate``) is
+     capitalized. Choose the name carefully. Good names tell the user what a template is good for, or what architecture
+     it implements.
+
+   * A template is "registered" by using the :mod:`template <pennylane.template>` decorator ``@template``.
+     This allows us to record the queue of operations of a template, which is very useful for testing:
+
+     .. code-block:: python
+
+        with qml.utils.OperationRecorder() as rec:
+            MyNewTemplate(...)
+
+        list_of_gates = rec.queue
+
+   * Your template should have a positional argument called ``wires``, which takes an integer (single wire) or a list
+     of wire indices that the template acts on.
+
+   * Consider using the :func:`broadcasting <pennylane.templates.broadcasting>` function to make your code more concise.
+
+   * Write an extensive docstring that explains how to use the template. Include a sketch of the template (add the
+     file to the ``doc/_static/templates/<templ_type>/`` directory). You can also display a small usage example
+     at the beginning of the docstring.
+
+     At the end of the docstring, add a section starting with the ``.. UsageDetails::`` directive,
+     where you demonstrate with code examples how to use the templates with different
+     settings, for example varying the number of wires, explaining keyword arguments and special cases.
+
+   * Check the inputs to the template. You can use the functions provided in :mod:`utils <pennylane.templates.utils>`.
+     Don't forget that arguments may be passed by the user to the qnode as positional or keyword arguments, and
+     by using different interfaces (i.e., a input could be a ``numpy.ndarray`` or a list of
+     :class:`Variable <pennylane.variable.Variable>`, depending on how the user uses the template).
```

Comment:
I've just noticed that the pennylane.templates.utils module is not linked to on the templates code page (https://pennylane.readthedocs.io/en/stable/code/qml_templates.html). It might be good to include it (somewhere) so that the documentation is available (as it is, I would have no idea what is available/how they're used without looking in the source code directly)

Reply:
Good point. Do you know how to use ``autosummary`` so that all functions from a file are listed (without that file being a module?)

Reply:
Yep! See https://pennylane.readthedocs.io/en/stable/development/guide.html#adding-a-new-module-to-the-docs

Reply:
Lol that was in the file I was adding to. I should have read it!

Reply:
I see, ``automodapi`` is what I tried before, and it does not import the functions. From playing around I assume that this is because they are private (i.e. they start with underscore). Since developers use them they are kind of semi-private.

Shall I refactor them to be public? Or resolve both a separate PR?

Reply:
Alternatively, I could list all functions manually by using ``autosummary``

Reply:
Ok, I solved it for now by using ``automodapi``, but the list in the docs (https://pennylane.readthedocs.io/en/stable/code/qml_templates.html) will be empty for now. I am preparing a separate PR that refactors the utils functions to be non-private. This is consistent with other utils modules in PennyLane...